File: feedfact/feedfact.rb
n, k, m, r = gets.split.map { |x| x.to_i }

csps = []
n.times do
    csps.push gets.split.map { |x| x.to_i }
end

row = { 0 => 0 }
row.default = 0
csps.each do |cost, prod|
    row.dup.each do |total_cost, max_prod|
        new_cost = total_cost + cost
        row[new_cost] = [row[new_cost], max_prod + prod].max
    end
end

es = row.to_a.map do |cost, prod|
         m.to_f - cost.to_f / prod.to_f unless cost > r || prod < k
     end.compact

best = es.max
if best and best > 0
    puts (best * 1000).truncate
else 
    puts -1
end

File: feedfact/feedfact.py
n, k, m, r = (int(x) for x in raw_input().split())
cs = []
ps = []
for i in range(n):
    c, p = (int(x) for x in raw_input().split())
    cs.append(c)
    ps.append(p)

row = {0: 0}
for cost, prod in zip(cs, ps):
    for total_cost, max_prod in row.copy().iteritems():
        row[total_cost + cost] = max(row.get(total_cost + cost, 0), max_prod + prod)


es = [m - float(cost) / float(prod) for cost, prod in row.iteritems() if cost <= r and prod >= k]

try:
    best = max(es)
except:
    print -1
if best > 0:
    print int(best * 1000)
else:
    print -1
